src/tscpp/util/unit_tests/test_Bravo.cc

```diff
@@ -0,0 +1,192 @@
+    with the License.  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+ */
+
+#include "catch.hpp"
+#include "tscpp/util/Bravo.h"
+
+#include <chrono>
+#include <mutex>
+#include <shared_mutex>
+#include <thread>
+
+using namespace std::chrono_literals;
+
+TEST_CASE("BRAVO - simple check with try-lock", "[libts][BRAVO]")
+{
+  SECTION("reader-reader")
+  {
+    ts::bravo::shared_mutex mutex;
+    ts::bravo::shared_lock<ts::bravo::shared_mutex> lock(mutex);
+    CHECK(lock.owns_lock() == true);
+
+    std::thread t{[](ts::bravo::shared_mutex &mutex) {
+                    ts::bravo::Token token{0};
+                    CHECK(mutex.try_lock_shared(token) == true);
+                    mutex.unlock_shared(token);
+                  },
+                  std::ref(mutex)};
+
+    t.join();
+  }
+
+  SECTION("reader-writer")
+  {
+    ts::bravo::shared_mutex mutex;
+    ts::bravo::shared_lock<ts::bravo::shared_mutex> lock(mutex);
+    CHECK(lock.owns_lock() == true);
+
+    std::thread t{[](ts::bravo::shared_mutex &mutex) { CHECK(mutex.try_lock() == false); }, std::ref(mutex)};
+
+    t.join();
+  }
+
+  SECTION("writer-reader")
+  {
+    ts::bravo::shared_mutex mutex;
+    std::lock_guard<ts::bravo::shared_mutex> lock(mutex);
+
+    std::thread t{[](ts::bravo::shared_mutex &mutex) {
+                    ts::bravo::Token token{0};
+                    CHECK(mutex.try_lock_shared(token) == false);
+                    CHECK(token == 0);
+                  },
+                  std::ref(mutex)};
+
+    t.join();
+  }
+
+  SECTION("writer-writer")
+  {
+    ts::bravo::shared_mutex mutex;
+    std::lock_guard<ts::bravo::shared_mutex> lock(mutex);
+
+    std::thread t{[](ts::bravo::shared_mutex &mutex) { CHECK(mutex.try_lock() == false); }, std::ref(mutex)};
+
+    t.join();
+  }
+}
+
+TEST_CASE("BRAVO - check with race", "[libts][BRAVO]")
+{
+  SECTION("reader-reader")
+  {
+    ts::bravo::shared_mutex mutex;
+    int i = 0;
+
+    std::thread t1{[&](ts::bravo::shared_mutex &mutex) {
+                     ts::bravo::shared_lock<ts::bravo::shared_mutex> lock(mutex);
+                     CHECK(lock.owns_lock() == true);
+                     ++i;
+                   },
+                   std::ref(mutex)};
+
+    std::thread t2{[&](ts::bravo::shared_mutex &mutex) {
+                     ts::bravo::shared_lock<ts::bravo::shared_mutex> lock(mutex);
+                     CHECK(lock.owns_lock() == true);
+                     ++i;
+                   },
+                   std::ref(mutex)};
+
+    t1.join();
+    t2.join();
+
+    CHECK(i == 2);
```

Comment:
Hmm why take a shared lock to write?  Seems like `i` could end up being 1 not 2.

Reply:
I'll get rid of this. Initially, incrementing `i` to prevent optimization and do nothing. But now it has some checks.